### PR TITLE
[15212] hotfix of BasicConfigurationExample cli

### PR DESCRIPTION
**Pull request by MiguelBarro**
## Description

Fixing the specific use case where the first argument is a flag (like `--transient`).
In that case, the mechanism that checks if the first argument is `publisher\subscriber` failed.

The changes are piggybacked to `DDSHelloWorldExample` that uses the same parsing strategy.

## Contributor Checklist
- [x] Commit messages follow the project guidelines. 
- [x] The code follows the style guidelines of this project. 
- [x] Tests that thoroughly check the new feature have been added/Regression tests checking the bug and its fix have been added. 
- [x] Any new/modified methods have been properly documented using Doxygen. 
- [x] Fast DDS test suite has been run locally. 
- [x] Changes are ABI compatible. 
- [x] Changes are API compatible. 
- [x] Documentation builds and tests pass locally. 
- [x] New feature has been added to the `versions.md` file (if applicable).
- [x] New feature has been documented/Current behavior is correctly described in the documentation. 

## Reviewer Checklist
- [x] Check contributor checklist is correct.
- [x] Check CI results: changes do not issue any warning.
- [x] Check CI results: failing tests are unrelated with the changes.

### CODE REVIEW

File: cmake/common/gtest.cmake [hunk 1]
```diff
@@ -24,7 +24,7 @@ macro(add_gtest)
         set(test "${ARGV0}")
         set(command "${test}")
     endif()
-    set(multiValueArgs SOURCES ENVIRONMENTS DEPENDENCIES LABELS)
+    set(multiValueArgs SOURCES ENVIRONMENTS DEPENDENCIES LABELS IGNORE)
```
Comment (jparisu) [inline cmake/common/gtest.cmake:27]: Not sure this was intented for this PR. Please check.

File: test/CMakeLists.txt [hunk 2]
```diff
@@ -25,6 +25,15 @@ endif()
 file(TO_CMAKE_PATH "${PROJECT_SOURCE_DIR}/valgrind.supp" MEMORYCHECK_SUPPRESSIONS_FILE_TMP)
 set(MEMORYCHECK_SUPPRESSIONS_FILE ${MEMORYCHECK_SUPPRESSIONS_FILE_TMP} CACHE FILEPATH "Valgrind suppresions file")
 
+# Check if /etc/hosts file has been modified to add dummy host test subjects
+if(WIN32)
+    execute_process(COMMAND powershell -C Resolve-DNSName www.acme.com.test
```
Comment (jparisu) [inline test/CMakeLists.txt:30]: Not sure this was intented for this PR. Please check.

File: test/unittest/utils/CMakeLists.txt [hunk 3]
```diff
@@ -111,7 +111,11 @@ target_include_directories(LocatorTests PRIVATE
     ${PROJECT_SOURCE_DIR}/include ${PROJECT_BINARY_DIR}/include ${Asio_INCLUDE_DIR}
     ${PROJECT_SOURCE_DIR}/src/cpp)
 target_link_libraries(LocatorTests GTest::gtest ${MOCKS})
-add_gtest(LocatorTests SOURCES ${LOCATORTESTS_SOURCE})
+if(EPROSIMA_TEST_DNS_NOT_SET_UP)
```
Comment (jparisu) [inline test/unittest/utils/CMakeLists.txt:114]: Not sure this was intented for this PR. Please check.

File: examples/cpp/dds/BasicConfigurationExample/BasicConfigurationSubscriber.cpp [hunk 4]
```diff
@@ -182,6 +184,8 @@ bool HelloWorldSubscriber::init(
     if (transient)
     {
         rqos.durability().kind = TRANSIENT_LOCAL_DURABILITY_QOS;
+        // mimic publisher behaviour. Assure the user see all samples are received
+        rqos.history().kind = KEEP_ALL_HISTORY_QOS;
```
Comment (jparisu) [inline examples/cpp/dds/BasicConfigurationExample/BasicConfigurationSubscriber.cpp:188]: If set with both reliability and durability, please add it also to Publisher

**Review by jparisu — CHANGES_REQUESTED** [review]

File: cmake/common/gtest.cmake (hunk 1 above)
Comment (MiguelBarro) [reply cmake/common/gtest.cmake]: Sorry, I departed from master but it was not aligned with the remote master branch. I've rebased already.

File: test/CMakeLists.txt (hunk 2 above)
Comment (MiguelBarro) [reply test/CMakeLists.txt]: see above

File: test/unittest/utils/CMakeLists.txt (hunk 3 above)
Comment (MiguelBarro) [reply test/unittest/utils/CMakeLists.txt]: see above.

File: examples/cpp/dds/BasicConfigurationExample/BasicConfigurationSubscriber.cpp (hunk 4 above)
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfigurationSubscriber.cpp:189]: done

**Review by jparisu — DISMISSED** [review]

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]
Please check numerical arguments are not negative.

File: examples/cpp/dds/BasicConfigurationExample/arg_configuration.h [hunk 5]
```diff
@@ -71,18 +73,48 @@ struct Arg : public option::Arg
             bool msg)
     {
         char* endptr = 0;
-        if (option.arg != 0 && strtol(option.arg, &endptr, 10))
+        if ( option.arg != nullptr)
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/BasicConfigurationExample/arg_configuration.h:76]:
```suggestion
        if ( option.arg != nullptr )
```

File: examples/cpp/dds/BasicConfigurationExample/arg_configuration.h [hunk 6]
```diff
@@ -170,10 +202,10 @@ const option::Descriptor usage[] = {
     { UNKNOWN_OPT, 0, "", "",                Arg::None,      "\nSubscriber options:"},
     { TOPIC, 0, "t", "topic",                  Arg::String,
       "  -t <topic_name> \t--topic=<topic_name>  \tTopic name (Default: HelloWorldTopic)." },
-    { DOMAIN_ID, 0, "d", "domain",                Arg::Numeric,
+    { DOMAIN_ID, 0, "d", "domain",                Arg::NumericRange<0, 230>,
       "  -d <id> \t--domain=<id>  \tDDS domain ID (Default: 0)." },
-    { SAMPLES, 0, "s", "samples",              Arg::Numeric,
-      "  -s <num> \t--samples=<num>  \tNumber of samples to wait for (Default: 0 => infinite samples)." },
+    { SAMPLES, 0, "s", "samples",              Arg::NumericRange<>,
+      "  -s <num> \t--samples=<num>  \tNumber of samples to send (Default: 0 => infinite samples)." },
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/BasicConfigurationExample/arg_configuration.h:208]:
```suggestion
      "  -s <num> \t--samples=<num>  \tNumber of samples to wait for (Default: 0 => infinite samples)." },
```

File: examples/cpp/dds/HelloWorldExample/HelloWorld_main.cpp [hunk 7]
```diff
@@ -73,18 +76,48 @@ struct Arg : public option::Arg
             bool msg)
     {
         char* endptr = 0;
-        if (option.arg != 0 && strtol(option.arg, &endptr, 10))
+        if ( option.arg != nullptr)
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/HelloWorldExample/HelloWorld_main.cpp:79]:
```suggestion
        if ( option.arg != nullptr )
```

File: examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp [hunk 8]
```diff
@@ -192,6 +192,7 @@ bool HelloWorldPublisher::init(
     if (transient)
     {
         wqos.durability().kind = TRANSIENT_LOCAL_DURABILITY_QOS;
+        wqos.reliability().kind = RELIABLE_RELIABILITY_QOS;
```
Comment (juanlofer-eprosima) [inline examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp:195]: I would not do this. But if set, the warning after parsing should be removed.

**Review by juanlofer-eprosima — CHANGES_REQUESTED** [review]

File: examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp (hunk 8 above)
Comment (juanlofer-eprosima) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp:195]: The warning of TLS hops being greater than 255 can also be omitted now.

File: examples/cpp/dds/BasicConfigurationExample/arg_configuration.h (hunk 5 above)
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/arg_configuration.h:76]: done

File: examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp (hunk 8 above)
Comment (MiguelBarro) [reply examples/cpp/dds/BasicConfigurationExample/BasicConfigurationPublisher.cpp:195]: fair enought

**Comment by MiguelBarro** [comment]
@richiprosima Please test this for me 🤯

**Review by juanlofer-eprosima — APPROVED** [review]
LGTM

**Comment by EduPonz** [comment]
@richiprosima please test this

**Comment by MRicoIE2C** [comment]
LGTM

**Pull request merged**